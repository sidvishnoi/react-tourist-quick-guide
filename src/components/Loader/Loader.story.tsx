import { color, number, select, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';

import * as React from 'react';
import Loader from '.';

const stories = storiesOf('Atoms/Loader', module);
stories.addDecorator(withKnobs);

stories.add('loader', () => {
  const props = {
    color: color('color', '#ff8a00'),
    size: number('size', 50, {
      max: 300,
      min: 20,
      range: true,
      step: 5,
    }),
    speed: select('speed', ['slow', 'normal', 'fast'], 'normal'),
  };
  return <Loader {...props} />;
});
