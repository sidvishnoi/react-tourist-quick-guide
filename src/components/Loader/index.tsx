import * as React from 'react';

interface LoaderProps {
  size?: number;
  color?: string;
  speed?: 'fast' | 'slow' | 'normal';
}

const enum Duration {
  slow = 1.5,
  normal = 1,
  fast = 0.5,
}

export default function Loader(props: LoaderProps) {
  const { size = 50, color = '#ff8a00', speed = 'normal' } = props;
  const duration =
    speed === 'normal'
      ? Duration.normal
      : speed === 'fast'
      ? Duration.fast
      : Duration.slow;
  // from: https://samherbert.net/svg-loaders/
  return (
    <svg
      width={size.toString()}
      height={size.toString()}
      viewBox="0 0 38 38"
      xmlns="http://www.w3.org/2000/svg"
      stroke={color}
      className="loader"
    >
      <g fill="none" fillRule="evenodd">
        <g transform="translate(1 1)" strokeWidth="2">
          <circle strokeOpacity=".5" cx="18" cy="18" r="18" />
          <path d="M36 18c0-9.94-8.06-18-18-18">
            <animateTransform
              attributeName="transform"
              type="rotate"
              from="0 18 18"
              to="360 18 18"
              dur={`${duration}s`}
              repeatCount="indefinite"
            />
          </path>
        </g>
      </g>
    </svg>
  );
}
