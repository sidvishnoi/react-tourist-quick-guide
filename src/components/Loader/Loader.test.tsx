import React = require('react');
import * as renderer from 'react-test-renderer';

import Loader from '.';

describe('Component - Loader', () => {
  it('matches snapshot - default', () => {
    const tree = renderer.create(<Loader />);
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot - option overrides', () => {
    const tree = renderer.create(
      <Loader size={100} color="red" speed="fast" />,
    );
    expect(tree).toMatchSnapshot();
  });
});
