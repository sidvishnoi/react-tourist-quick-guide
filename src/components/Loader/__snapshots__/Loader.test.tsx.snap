// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Loader matches snapshot - default 1`] = `
<svg
  className="loader"
  height="50"
  stroke="#ff8a00"
  viewBox="0 0 38 38"
  width="50"
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    fill="none"
    fillRule="evenodd"
  >
    <g
      strokeWidth="2"
      transform="translate(1 1)"
    >
      <circle
        cx="18"
        cy="18"
        r="18"
        strokeOpacity=".5"
      />
      <path
        d="M36 18c0-9.94-8.06-18-18-18"
      >
        <animateTransform
          attributeName="transform"
          dur="1s"
          from="0 18 18"
          repeatCount="indefinite"
          to="360 18 18"
          type="rotate"
        />
      </path>
    </g>
  </g>
</svg>
`;

exports[`Component - Loader matches snapshot - option overrides 1`] = `
<svg
  className="loader"
  height="100"
  stroke="red"
  viewBox="0 0 38 38"
  width="100"
  xmlns="http://www.w3.org/2000/svg"
>
  <g
    fill="none"
    fillRule="evenodd"
  >
    <g
      strokeWidth="2"
      transform="translate(1 1)"
    >
      <circle
        cx="18"
        cy="18"
        r="18"
        strokeOpacity=".5"
      />
      <path
        d="M36 18c0-9.94-8.06-18-18-18"
      >
        <animateTransform
          attributeName="transform"
          dur="0.5s"
          from="0 18 18"
          repeatCount="indefinite"
          to="360 18 18"
          type="rotate"
        />
      </path>
    </g>
  </g>
</svg>
`;
